### DISCUSSION TITLE
Trust Fund Workshop Doesn't Work!

### DISCUSSION THREAD

**Discussion by JamesLampert** [reach-sh/reach-lang#1545] [Q&A]
I started on the Trust Fund Workshop last year. Real life intervened, along with the new RSVP Tutorial (which still has a few things I'm having trouble wrapping my mind around), and I only got back to Trust Fund this month.

And the version I left off with had (and still has) a problem with the "Third Party" ("Bystander" in the canned solution) always getting the money.

This afternoon, I copied and pasted the "canned solution" from the instruction page into a workspace -- _and IT has the SAME BEHAVIOR!_

So I was tearing my hair out, trying to get my version to work, trying to find a substantive difference from the canned solution (other than the fact that mine doesn't have the "Give Chance" function), and now I find that I was searching in vain, because the canned solution doesn't work either.

Can somebody make some sense of this?

See the thread, "Trust Fund Workshop Weirdness" for my earlier observations (and a link to a "gist" containing my version).

**Comment by jeapostrophe** [comment]
Can you move this to that other thread (which I can't find) so we have the context? I vaguely remember talking about this and giving an explanation, but I don't recall how things were left and this message doesn't say what you think the problem is.

**Reply by JamesLampert** [reply]
I didn't create any "other thread" about this in GitHub about this. But I did create a fair amount of Discord traffic.

I've been going through my postings in the original "Help" discussion on Discord, and simultaneously checking the creation dates on my Reach workspaces, and it seems I misremembered some things.

First (but after doing the original RPS tutorial under Ms. Hodge's video tutelage, and Hash-Lock, and Relay), I did the Trust Fund workshop in August. And it evidently worked as advertised, back in August. Then (according to my Discord traffic, although I can't find the workspace), I did the "Fair RPS" workshop (without ever finding a solution that satisfied my notion of "fair"), and the FOMO workshop. Then, finding out that the RSVP tutorial had just been added (I used to work with a guy who was a big "Buffy the Vampire Slayer" fan; because of his influence, I still tend to name buffers "Buffy), I went through it and had to step away and do "real work for real customers" for quite some time. Then, earlier this month, I began running through RSVP again, to see how far I'd gotten, and then to get as much understanding of the new concepts it introduced as possible. And I posted some nagging unanswered questions about it in a thread titled, "Not sure I understand this," in "help-forum" on Discord, https://discord.com/channels/628402598663290882/1063616294194262147

Then, I went back to Trust Fund, in hopes of attempting "Guardian Account," which would put me on track to the real-world application my boss is after. And I found out that Trust Fund was now unconditionally giving the money to "Third Party." And after tearing my hair out, trying to figure out how my solution differed from the "canned solution," I created another workspace, and copied and pasted the "canned solution" into it.

My trials and tribulations with Trust Fund are detailed in the "Trust Fund Workshop weirdness" thread of "help-forum" on Discord, https://discord.com/channels/628402598663290882/1064699938950348940
which contains an extended exchange between me and "BMscis," who tried my Trust Fund solution, and found that it worked perfectly for him (her? them? it?).

And the latest development is that I tried doing a ./reach update, and found that I was once again in the situation of update failing.

**Reply by jeapostrophe** [reply]
Okay, so you're saying that you think Trust Fund is working incorrectly because it is "unconditionally giving the money to "Third Party."".

In the sample code --- https://github.com/reach-sh/reach-lang/blob/master/examples/workshop-trust-fund/index.mjs#L59-L61 --- we run three times, each time delaying a different participant. When I run this, I get output like:

https://gist.githubusercontent.com/jeapostrophe/1644158249524924122ca44f12376b63/raw/7fa384a7b4d995e69793e52132a6e84fd7ca34d4/ALGO.workshop-trust.fund

If you search for "has a balance of", you can see the main part of the tests.

In the first run, we see:
```
Funder has a balance of 89.996
Receiver has a balance of 109.997
Bystander has a balance of 100
```

In the second, it is:
```
Funder has a balance of 89.996
Receiver has a balance of 100.0001
Bystander has a balance of 109.9971
```

And in the third it is:
```
Funder has a balance of 89.9961
Receiver has a balance of 100.0002
Bystander has a balance of 109.9972
```

**Reply by JamesLampert** [reply]
In a typical run for me, using the "canned solution," I get (as shown in my Discord thread):

First run:
```
Funder has a balance of 89.9967
Receiver has a balance of 100.0008
Bystander has a balance of 109.9978
```

Second run:
```
Funder has a balance of 89.9968
Receiver has a balance of 100.0009
Bystander has a balance of 109.9979
```

Third run:
```
Funder has a balance of 89.9972
Receiver has a balance of 100.0013
Bystander has a balance of 109.9983
```
and using my solution (gist https://gist.github.com/JamesLampert/13aec95ef996cc947f1b3e62702db733), I get:

First run:
```
Funder ends up with a balance of 89.9967
Receiver ends up with a balance of 100.0008
Third Party ends up with a balance of 109.9978
```

Second run:
```
Funder ends up with a balance of 89.9966
Receiver ends up with a balance of 100.0007
Third Party ends up with a balance of 109.9977
```

Third Run:
```
Funder ends up with a balance of 89.997
Receiver ends up with a balance of 100.0011
Third Party ends up with a balance of 109.9981
```

Using my code, from my gist, "BMscis" gets substantially the same result you got with the canned solution.

But now, I'm now starting to notice something else: _neither_ your results, _nor_ mine, _nor_ those of "BMscis" agree with the intended results, as given in the instructions:

First run (goes to Receiver):
```
Funder has a balance of 89.99926093
Receiver has a balance of 109.999956574
Bystander has a balance of 100.0
```

Second run (goes back to Funder):
```
Funder has a balance of 99.999217452
Receiver has a balance of 99.99995488
Bystander has a balance of 100.0
```

Third run (goes to Bystander):
```
Funder has a balance of 89.99921581
Receiver has a balance of 100.0
Bystander has a balance of 109.999956956
```

Something is clearly weird here.

**Reply by JamesLampert** [reply]
I've posted the details of several iterations of trying to make this workshop behave as advertised to the Reach Discord help-forum thread titled "Trust Fund Workshop Weirdness."

Those results can be summed up very simply: I and several others have run the "canned solution" to the trust fund workshop, and (now that I'm on a current version of Reach), my results agree with what the others have gotten: the second pass, which is supposed to have the algos returned to the funder, instead sends them to the "bystander."  I have tried modifying the canned solution, and whether I shorten the funder wait time, lengthen the funder wait time, or add a bystander wait time, I cannot find a way to force algos to return to the Funder.
